### DISCUSSION TITLE
Enable Link-Time Optimization (LTO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [tramhao/termusic#372] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

I found that LTO was [removed](https://github.com/tramhao/termusic/commit/24f40f6e56c8554e5206d88e72a400ed2c8c0a8c) from the project for some reason but I didn't find that reason. I also read [this](https://github.com/tramhao/termusic/issues/149) thread about crashes with LTO. I have seen several similar reports about pkgbuild and its LTO. According to my findings and experiments, it has nothing from the Cargo's LTO side - it's something strange in pkgbuild settings with LTO. So I think it should be pretty safe to enable LTO here.

Thank you.

**Comment by tramhao** [comment]
This could be turned on. I saw some other optimization flags also. Not sure what is best for us.

**Comment by hasezoey** [comment]
I dont know why lto was actually disabled, but my guesses are:
- on arch LTO had a problem, see #149, though i think this is unrelated to the rust LTO
- before 0.8(yanked)/0.9.0, the `debug` profile was basically broken, so i assume it was disabled to speed-up compile time, see #158

i personally had tried enabling rust's LTO and not had a problem, i will put in a PR to re-enable it for the next version and see what happens from there

EDIT: suggested with #380
